### ISSUE TITLE
Update and discourage use of controller specs

### ISSUE THREAD

**Issue by corroded** [rubocop-hq/rspec-style-guide#113]
The guide still has a section for controller specs. I think it should be mentioned that they have been deprecated and discouraged since Rails 3 (or was it 4?). I reckon the guidelines can *probably* stay for legacy stuff, but a suggestion mentioning request specs might be a good idea.

**Comment by pirj** [comment]
Absolutely. The guide is behind the reality in this regard. There's a mention of request specs here https://github.com/rubocop-hq/rspec-style-guide/#rails-integration, but the problem with slow request specs [is long gone](https://www.youtube.com/watch?v=oT74HLvDo_A).

I'm against of keeping controller spec guidelines and recommending request specs over controller specs at the same time. It's time for the change.

Would you like to take a stab at it?

Useful resources:
 - https://relishapp.com/rspec/rspec-rails/v/4-0/docs/request-specs/request-spec
 - https://guides.rubyonrails.org/testing.html#integration-testing
 - Rails' reasoning why controller specs are deprecated

**Comment by andyw8** [comment]
For reference, it was Rails 5 when controller tests were deprecated:

https://blog.bigbinary.com/2016/04/19/changes-to-test-controllers-in-rails-5.html

**Comment by pirj** [comment]
Also this https://github.com/rails/rails/issues/22496

**Comment by rpbaptist** [comment]
> For reference, it was Rails 5 when controller tests were deprecated:

I think this is an overstatement leading to current misunderstandings. `ActionController::TestCase` is deprecated, which includes `assigns` and `assert_template`. There's more to controller testing than those methods.

From the blog post you linked:

> According to Rails team, controller tests should be more concerned about what is the result of that controller action like what cookies are set, or what HTTP code is set rather than testing of the internals of the controller. So, these methods are removed from the core.

From the original issue: https://github.com/rails/rails/issues/22076

> ActionController::TestCase is deprecated. Let's get the controller test generation using our ActionDispatch::IntegrationTest future instead.

So controller tests are still around. It's just not testing which templates are rendered and which instance variables are set.

**Comment by pirj** [comment]
Agree, functional controller tests (that map to RSpec Rails' Controller specs) are said to [use `ActionDispatch::IntegrationTest` in Rails's testing guide](https://guides.rubyonrails.org/testing.html#functional-tests-for-your-controllers), not [`ActionController::TestCase::Behavior` as RSpec Rails states](https://relishapp.com/rspec/rspec-rails/v/4-0/docs/controller-specs).

This needs to be fixed in RSpec Rails' docs and reflected in the guideline as well.
It's just those two, `asssigns` and `assert_template` that were removed/extracted:

["Removed `assigns` and `assert_template`. Both methods have been extracted into the [`rails-controller-testing`](https://github.com/rails/rails-controller-testing) gem [(Pull Request)](https://github.com/rails/rails/pull/20138)"](https://github.com/rails/rails/blob/eca6c273fe2729b9634907562c2717cf86443b6b/guides/source/5_0_release_notes.md#L319).

**Comment by pirj** [comment]
https://discuss.rubyonrails.org/t/need-controller-side-effects-testing-advise/76103

**Comment by klyonrad** [comment]
So wait a second, either there is some massive hair splitting going on or all this *controller vs request specs* thing is based on misunderstandings.

What is supposed to be the difference between controller tests and integration tests in a rails non-rspec test suite?

They both are using the same class
https://github.com/rails/rails/blob/master/railties/lib/rails/generators/test_unit/controller/templates/functional_test.rb.tt
https://github.com/rails/rails/blob/master/railties/lib/rails/generators/test_unit/integration/templates/integration_test.rb.tt
